Add support for unsized types (#218)
minirust/minirust: spec/lang/representation.md

This is the *[representation relation]*, which is defined in the following.
`decode` converts a list of bytes into a value; this operation can fail if the byte list is not a valid encoding for the given type.
`encode` inverts `decode`; it will always work when the value is [well-formed][well-formed-value] for the given type (which the specification must ensure, i.e. violating this property is a spec bug).

> Unsized types cannot be represented as values and thus a call to `encode` or `decode` in a well formed program is a spec bug.

Reviewer: ```suggestion
Unsized types cannot be represented as values and thus a call to `encode` or `decode` for such types can never occur in a well-formed program.
```

    let bytes_data = M::T::ENDIANNESS.encode(Unsigned, M::T::PTR_SIZE, ptr.addr).unwrap();
    bytes_data.map(|b| AbstractByte::Init(b, ptr.provenance))
    
}

fn encode_ptr_meta<M: Memory>(meta: Option<PointerMeta>) -> List<AbstractByte<M::Provenance>> {
    match meta {
        None => list!(),
        Some(PointerMeta::ElementCount(len)) => {
            let bytes = M::T::ENDIANNESS.encode(Unsigned, M::T::PTR_SIZE, len).unwrap();
            bytes.map(|b| AbstractByte::Init(b, None))
        }
    }
}

impl Type {
    fn decode<M: Memory>(Type::Ptr(ptr_type): Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
        if bytes.len() < M::T::PTR_SIZE.bytes() { throw!(); }
        let ptr = decode_ptr::<M>(bytes.subslice_with_length(Int::ZERO, M::T::PTR_SIZE.bytes()))?;
        if !ptr_type.addr_valid(ptr.addr) {
            return None;
        }
        // TODO: Decode wide pointers, but needs slight changes in PtrType & Layout
        ret(Value::Ptr(ptr.widen(None)))
    }
    fn encode<M: Memory>(Type::Ptr(_): Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
        let Value::Ptr(ptr) = val else { panic!() };

>         encode_ptr::<M>(ptr.thin_pointer).iter().chain(encode_ptr_meta::<M>(ptr.metadata).iter()).collect()

Reviewer: Please let-bind the two components, so that it's not one line doing too many things.

```rust
impl<Provenance> DefinedRelation for Pointer<Provenance> {
    fn le_defined(self, other: Self) -> bool {
        self.thin_pointer.addr == other.thin_pointer.addr &&
        // Feedback wanted: Using `Option<PointerMeta> as DefinedRelation` would be okay aswell, 
        // since all antecedents garantee that either both or none have metadata in the first place?

>         self.metadata == other.metadata && 

Reviewer: In the future PointerMeta will contain vtable pointers that can have provenance, so it's probably better to define a `DefinedRelation` on `PointerMeta`.
